File: mlmc/mlmc_test.py
import math
import time
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.gridspec  as gridspec
from mlmc.mlmc import mlmc

        

def mlmc_test(mlmc_fn, M, N, L, N0, Eps, nvert, validate=False, validation_value=None, **mlmc_l_kwargs):
    """
    Runs a MLMC test for a given function mlmc_fn, number of levels L,
    number of samples N, initial sample size N0, and a list of desired accuracies Eps.
    The function mlmc_fn should take two arguments: level l and number of samples N.
    It should return two arrays: sum1 and sum2, where sum1 contains the first four moments
    of the fine level and sum2 contains the first two moments of the coarse level.
    The function will plot the results and return final mlmc estimate.
    """
    del1 = []; del2 = [] #del1 = P_l - P_{l-1}, del2 = P_l
    var1 = []; var2 = [] #var1 = var(P_l - P_{l-1}), var2 = var(P_l)
    kur1 = []
    chk1 = []
    cost = [] # cost of each level
    L = np.arange(0, L+1)

    for l in L:
        print(f"l = {l}")
        start_time = time.time()
        sum1, sum2 = mlmc_fn(l, N, **mlmc_l_kwargs)
        time_taken = time.time() - start_time
        cost.append(time_taken)
        sum1 = sum1 / N
        sum2 = sum2 / N
        # kurtosis
        numerator = sum1[3]- 4 * sum1[2] * sum1[0]+ 6 * sum1[1] * sum1[0]**2- 3 * sum1[0]**4
        denominator = (sum1[1] - sum1[0]**2)**2
        kurt = numerator / denominator if denominator > 0 else np.nan

        del1.append(sum1[0])
        del2.append(sum2[0])
        var1.append(sum1[1] - sum1[0]**2)
        var = max(sum2[1] - sum2[0]**2, 1e-12) #handles var=0 cases
        var2.append(var)
        kur1.append(kurt)

        if l==0:
            check = 0
        else:
            check = abs(del1[l] + del2[l-1] - del2[l]) /  \
                (3.0 * (np.sqrt(var1[l])) + np.sqrt(var2[l-1]) + np.sqrt(var2[l])/np.sqrt(N))
        chk1.append(check)

    # convert items to numpy arrays so I can do index selecting
    del1 = np.array(del1)
    del2 = np.array(del2)
    var1 = np.array(var1)
    var2 = np.array(var2)
    kur1 = np.array(kur1)
    chk1 = np.array(chk1)

    # use linear regression to estimate alpha, beta and gamma
    start = max(2, int(np.floor(0.4 * len(L))))
    range_ = np.arange(start, len(L))

    print('Estimates of key MLMC Theorem parameters based on linear regression: ')

    pa = np.polyfit(L[range_], np.log2(abs(del1[range_])), 1)
    alpha = -pa[0]
    print(f"alpha = {alpha:.6f} (exponent for MLMC weak convergence)")

    pb = np.polyfit(L[range_], np.log2(abs(var1[range_])), 1)
    beta = -pb[0]
    print(f"beta = {beta:.6f} (exponent for MLMC variance)")

    gamma = np.log2(cost[-1]/cost[-2])
    print(f"gamma = {gamma:.6f} (exponent for MLMC cost")

    # Perform checks on consistency error and kurtosis
    if max(chk1) > 1:
        print(f"WARNING: maximum consistency error = {max(chk1):.6f}")
        print(f"Indicates identitiy E[Pf-Pc] = E[Pf] - E[Pc] not satisfied")

    if kur1[-1] > 100:
        print(f"WARNING: kurtosis on finest level = {kur1[-1]:.6f}. \n \
                indicates MLMC correction dominated by a few rare paths; for information \
                on the connection to variance of sample variances see \
                http://mathworld.wolfram.com/SampleVarianceDistribution.html")

    # Plot figures
    nrows = nvert + 2 if validate else nvert + 1
    nrows = 5
    fig = plt.figure(figsize=(10, 3.5 * nrows))
    gs = gridspec.GridSpec(nrows, 2, figure=fig)
    axs = []
    for row in range(nrows - 2): # rows 0, 1, 2
        axs.append(fig.add_subplot(gs[row, 0]))
        axs.append(fig.add_subplot(gs[row, 1]))

    var_ax, mean_ax, chk_ax, kur_ax, count_ax, cost_ax = axs
    # MLMC and MC validation plot
    validation_ax = fig.add_subplot(gs[3, :])
    
    # Results plot
    results_ax = fig.add_subplot(gs[4, :])

    var_ax.plot(L, np.log2(var2), '-*', label='P_l')
    var_ax.plot(L[1:], np.log2(var1[1:]), '--*', label='P_l - P_{l-1}')
    var_ax.set_xlabel(r'level $l$')
    var_ax.set_ylabel(r'$\log_2$ variance')
    var_ax.legend(loc='upper right')

    mean_ax.plot(L, np.log2(np.abs(del2)), '-*', label='P_l')
    mean_ax.plot(L[1:], np.log2(np.abs(del1[1:])), '--*', label='P_l - P_{l-1}')
    mean_ax.set_xlabel(r'level $l$')
    mean_ax.set_ylabel(r'$\log_2 |\text{mean}|$')
    mean_ax.legend(loc='upper right')

    if nvert == 3:
        chk_ax.plot(L[1:] - 1e-9, chk1[1:], '--*')
        chk_ax.set_xlabel(r'level $l$')
        chk_ax.set_ylabel('consistency check')

        kur_ax.plot(L[1:] - 1e-9, kur1[1:], '--*')
        kur_ax.set_xlabel(r'level $l$')
        kur_ax.set_ylabel('kurtosis')

        # Plot target value of estimator
        if validation_value is not None:
            validation_ax.axhline(y=validation_value, linestyle='--', color='crimson', linewidth=2, label='Target QoI')

        # MLMC estimates
        mlmc_estimator = np.cumsum(del1)
        mlmc_se = np.sqrt(np.cumsum(var1) / N)
        validation_ax.errorbar(L, mlmc_estimator, yerr=mlmc_se, fmt='o-', capsize=5, elinewidth=1.2, color='blue', ecolor='lightblue',
                               label=r'MLMC estimates $\pm \sigma$ ')
        
        # Plot E[P_l] for context
        mc_se = np.sqrt(var2 / N)
        validation_ax.errorbar(L, del2, yerr=mc_se, fmt='s--', elinewidth=1.2, color='gray', ecolor='silver', label=r'MC estimates $\pm \sigma$')

        #Formatting
        validation_ax.set_xlabel(r'Level $\ell$', fontsize=12)
        validation_ax.set_ylabel(r'Estimate', fontsize=12)
        validation_ax.tick_params(axis='both', which='major', labelsize=10)
        validation_ax.grid(True, linestyle='--', linewidth=0.5, alpha=0.7)
        validation_ax.spines['top'].set_visible(False)
        validation_ax.spines['right'].set_visible(False)
        validation_ax.set_title('Convergence of MLMC Estimate', fontsize=13)
        validation_ax.legend(loc='best', frameon=True, fontsize=11)


    if nvert == 1:
        fig, axs = plt.subplots(1, 2, figsize=(8, 4.5))
    
    # Run MLMC for different EPS
    Nls = []
    ls = []
    maxl = 0
    mlmc_cost = []
    std_cost = []
    mlmc_solns = []

    for i, eps in enumerate(Eps):
        print(f"eps = {eps}")
        gamma = np.log2(M)

        # Run MLMC
        P, Nl = mlmc(N0, eps, mlmc_fn, alpha, beta, gamma, **mlmc_l_kwargs)
        l = len(Nl) - 1
        maxl = max(maxl, l) # See how many levels were needed to get our desired level of eps

        # Compute cost estimates - compares mlmc cost with standard mc cost
        levels = np.arange(0, l + 1)
        mlmc_c = (1 + 1 / M) * np.sum(Nl * M**levels) #C_tot = sum of Cl*Nl
        std_c = np.sum((2 * var2[-1] / eps**2) * M**levels) #var2 contains powers of P^f

        mlmc_cost.append(mlmc_c)
        std_cost.append(std_c)

        # Store per-level sample counts
        Nls.append(Nl)
        ls.append(levels)

        # Store mlmc solutions
        mlmc_solns.append(P)

    # Now make sure all the arrays we will be plotting are the same size
    for j in range(len(Eps)):
        current_Nl = Nls[j]
        current_ls = ls[j]

        pad_len = maxl + 1 - len(current_Nl)

        if pad_len > 0:
            Nls[j] = np.pad(current_Nl, (0, pad_len), constant_values=current_Nl[-1])
            ls[j] = np.pad(current_ls, (0, pad_len), constant_values=current_ls[-1])
    
    
    Nls_array = np.column_stack(Nls)  # shape: (maxl+1, len(Eps))
    ls_array = np.column_stack(ls)

    count_ax.set_prop_cycle(None)
    for i in range(len(Eps)):
        count_ax.semilogy(ls_array[:, i], Nls_array[:, i], '--o', label=f'$\\varepsilon$ = {Eps[i]:.3g}')
    count_ax.set_xlabel('level $\\ell$')
    count_ax.set_ylabel('$N_\\ell$')
    count_ax.legend()

    eps_squared = np.array(Eps) ** 2
    cost_ax.set_prop_cycle(None)
    cost_ax.loglog(Eps, eps_squared * std_cost, '--o', label='Std MC')
    cost_ax.loglog(Eps, eps_squared * mlmc_cost, '--x', label='MLMC')
    cost_ax.set_xlabel('accuracy $\\epsilon$')
    cost_ax.set_ylabel('$\\epsilon^2$ Cost')
    cost_ax.legend(loc='upper right')

    # arrange Eps in ascending order to show error getting smaller and smaller
    results_ax.plot(Eps, mlmc_solns, '--*', label='MLMC estimates')
    results_ax.fill_between(
        Eps,
        np.array(mlmc_solns) - np.array(Eps),
        np.array(mlmc_solns) + np.array(Eps),
        color='crimson',
        alpha=0.2,
        label=r'$\pm \epsilon^2$'
    )
    results_ax.set_xlabel(r'standard error, $\epsilon^2$')
    results_ax.set_ylabel(r'MLMC estimate')
    results_ax.invert_xaxis()
    if validation_value is not None:
        results_ax.axhline(y=validation_value, linestyle='--', color='crimson', linewidth=2, 
                           label='True QoI')
    results_ax.legend()

    plt.tight_layout()
    plt.show()

    return del1, del2, var1, var2

File: mlmc_cpp/src/mlmc_test.cpp

#include "mlmc_test.hpp"
#include "mlmc.hpp"
#include <iostream>
#include <vector>
#include <numeric>
#include <cmath>
#include <functional>
#include <algorithm>
#include <chrono>
#include <ctime>
#include <fstream>
#include <string>
#include <iomanip>
#include <limits>

void mlmc_test(
    std::function<std::pair<std::vector<double>, std::vector<double>>(int, int)> mlmc_fn,
    int M, int N, int L, int N0, const std::vector<double>& Eps,
    const std::string& output_convergence_filename, 
    const std::string& output_complexity_filename
) {
    std::vector<double> del1, del2, var1, var2, kur1, chk1, cost;
    std::vector<int> levels;

    // Print header with date/time
    time_t now = time(nullptr);
    char date[64];
    strftime(date, sizeof(date), "%c", localtime(&now));
    printf("\n**********************************************************\n");
    printf("*** MLMC file version 1.0     produced by              ***\n");
    printf("*** C++ mlmc_test on %s         ***\n", date);
    printf("**********************************************************\n");
    printf("\n**********************************************************\n");
    printf("*** Convergence tests, kurtosis, telescoping sum check ***\n");
    printf("*** using N =%7d samples                           ***\n", N);
    printf("**********************************************************\n");
    printf("\n l   ave(Pf-Pc)    ave(Pf)   var(Pf-Pc)    var(Pf)    kurtosis     check        cost \n");
    printf("--------------------------------------------------------------------------------------\n");

    for (int l = 0; l <= L; ++l) {
        levels.push_back(l);
        std::cout << "l = " << l << std::endl;
        auto start = std::chrono::high_resolution_clock::now();
        auto [sum1, sum2] = mlmc_fn(l, N);
        auto end = std::chrono::high_resolution_clock::now();
        std::chrono::duration<double> elapsed = end - start;
        cost.push_back(elapsed.count());
        
        // Estimator calculation
        for (auto& x : sum1) x /= N;
        for (auto& x : sum2) x /= N;
        
        // Kurtosis calculation
        double numerator = sum1[3] - 4 * sum1[2] * sum1[0] + 6 * sum1[1] * sum1[0] * sum1[0] - 3 * std::pow(sum1[0], 4);
        double denominator = std:: pow(sum1[1] - sum1[0] * sum1[0], 2);
        double kurt = (denominator > 0) ? numerator / denominator : NAN;

        // Add elements to vectors
        del1.push_back(sum1[0]);
        del2.push_back(sum2[0]);
        var1.push_back(sum1[1] - sum1[0] * sum1[0]);
        double var = std::max(sum2[1] - sum2[0] * sum2[0], 1e-12);
        var2.push_back(var);
        kur1.push_back(kurt);

        double check = 0.0;
        if (l > 0) {
            check = std::abs(del1[l] + del2[l-1] - del2[l]) / 
            (3.0 * std::sqrt(var1[l]) + std::sqrt(var2[l-1]) + std::sqrt(var2[l]) / std::sqrt(N));
        }
        chk1.push_back(check);

        // Print formatted results for this level
        printf("%2d  %11.4e %11.4e %11.4e %11.4e %11.4e %11.4e %11.4e\n",
            l, del1[l], del2[l], var1[l], var2[l], kur1[l], chk1[l], cost[l]);
        

    }

    // Print the MLMC estimator
    double estimator = std::accumulate(del1.begin(), del1.end(), 0.0);
    std::cout << "\nMLMC estimator = " << estimator << std::endl;

    // regression estimates
    int start = std::max(2, static_cast<int>(std::floor(0.4 * levels.size())));
    std::vector<int> reg_levels(levels.begin() + start, levels.end());
    std::vector<double> reg_del1(del1.begin() + start, del1.end());
    std::vector<double> reg_var1(var1.begin() + start, var1.end());

    std::cout << "\nEstimates of key MLMC Theorem parameters based on linear regression:\n";
    double alpha = regression(reg_levels, reg_del1);
    
    double beta = regression(reg_levels, reg_var1);
    std::cout << "beta = " << beta << " (exponent for MLMC variance)\n";

    double gamma = std::log2(cost.back() / cost[cost.size() - 2]);
    std::cout << "gamma = " << gamma << " (exponent for MLMC cost)\n";

    // Consistency error and kurtosis warnings
    double max_chk = *std::max_element(chk1.begin(), chk1.end());
    if (max_chk > 1.0) {
        std::cout << "WARNING: maximum consistency error = " << max_chk << "\n";
        std::cout << "Indicates identity E[Pf-Pc] = E[Pf] - E[Pc] not satisfied\n";
    }
    if (kur1.back() > 100.0) {
        std::cout << "WARNING: kurtosis on finest level = " << kur1.back() << "\n";
        std::cout << "Indicates MLMC correction dominated by a few rare paths; see http://mathworld.wolfram.com/SampleVarianceDistribution.html\n";
    }

    printf("\n******************************************************\n");
    printf("*** Linear regression estimates of MLMC parameters ***\n");
    printf("******************************************************\n");
    printf("\n alpha = %f  (exponent for MLMC weak convergence)\n", alpha);
    printf(" beta  = %f  (exponent for MLMC variance) \n", beta);
    printf(" gamma = %f  (exponent for MLMC cost) \n", gamma);

    // Output results to file
    std::ofstream file_out(output_convergence_filename);
    file_out << "level,ave_Pf-Pc,ave_Pf,var_Pf-Pc,var_Pf,kurtosis,check,cost,N\n";
    for (int i = 0; i <= L; ++i) {
        file_out << levels[i] << ","
                << del1[i] << ","
                << del2[i] << ","
                << var1[i] << ","
                << var2[i] << ","
                << kur1[i] << ","
                << chk1[i] << ","
                << cost[i] << ","
                << N << "\n";
    }
    file_out.close();
    std::cout << "\nWrote convergence table to " << output_convergence_filename << "\n";

    // Complexity tests: Running MLMC for different Eps values
    std::vector<std::vector<int>> Nls;
    std::vector<std::vector<double>> costs;
    std::vector<double> mlmc_estimates;
    std::vector<double> std_costs;
    std::vector<double> mlmc_costs;
    std::vector<std::vector<double>> Yls;  // Store per-level estimators for each epsilon
    std::vector<std::vector<double>> Vls; // Store per-level variances for each epsilon

    std::cout << "\n*********** Complexity test (MLMC vs standard MC for different eps) ***********\n";
    for (const auto& eps: Eps) {
        std::cout << "Running MLMC for eps = " << eps << "...\n";
        auto [P, Nl, suml] = mlmc(N0, eps, mlmc_fn, alpha, beta, gamma);

        mlmc_estimates.push_back(P);
        Nls.push_back(Nl);

        std::vector<double> means(Nl.size()), variances(Nl.size());
        for (size_t l = 0; l < Nl.size(); ++l) {
            if (Nl[l] > 0) {
                double mean = suml[0][l] / Nl[l];
                double mean_sq = suml[1][l] / Nl[l];
                means[l] = mean;
                variances[l] = std::max(0.0, mean_sq - mean * mean);
            } else {
                means[l] = 0.0;
                variances[l] = 0.0;
            }
        }
            
        Yls.push_back(means);
        Vls.push_back(variances);

        // Compute MLMC cost = sum_{l} Nl[l] * Cl[l]
        std::vector<double> Cl(Nl.size());
        for (size_t l = 0; l < Nl.size(); ++l)
            Cl[l] = std::pow(M, l);
        
        double mlmc_cost = 0.0;
        for (size_t l = 0; l < Nl.size(); ++l)
            mlmc_cost += Nl[l] * Cl[l];
        
        // Standard MC cost estimate: var2[-1] is the finest-level variance
        double var_fine = var2.back();
        double std_cost = 0.0;
        for (size_t l = 0; l < Nl.size(); ++l)
            std_cost += (2 * var_fine / (eps * eps)) * Cl[l];

        mlmc_costs.push_back(mlmc_cost);
        std_costs.push_back(std_cost);

        std::cout << " MLMC estimate = " << P << ", MLMC cost = " << mlmc_cost << ", Std MC cost = " << std_cost << std::endl;
    }

    // Output complexity data
    size_t max_levels = 0;
    for (const auto& Nl : Nls) max_levels = std::max(max_levels, Nl.size());

    std::ofstream comp_out(output_complexity_filename);
    comp_out << "eps,mlmc_estimate,mlmc_cost,std_mc_cost";
    for (size_t l = 0; l < Nls[0].size(); ++l)
        comp_out << ",Nl_" << l;
    for (size_t l = 0; l < Yls[0].size(); ++l) {
        comp_out << ",Yl_" << l;
    }
    for (size_t l = 0; l < Vls[0].size(); ++l) {
        comp_out << ",Vl_" << l;
    }
    comp_out << "\n";

    for (size_t i = 0; i < Eps.size(); ++i) {
        comp_out << Eps[i] << "," << mlmc_estimates[i] << "," << mlmc_costs[i] << "," << std_costs[i];

        // Pad with zeros if this run didn't have all levels
        for (size_t l = 0; l < max_levels; ++l)
            comp_out << "," << (l < Nls[i].size() ? Nls[i][l] : std::numeric_limits<double>::quiet_NaN());
        for (size_t l = 0; l < max_levels; ++l)
            comp_out << "," << (l < Yls[i].size() ? Yls[i][l] : std::numeric_limits<double>::quiet_NaN());
        for (size_t l = 0; l < max_levels; ++l)
            comp_out << "," << (l < Vls[i].size() ? Vls[i][l] : std::numeric_limits<double>::quiet_NaN());

        comp_out << "\n";
    }
    comp_out.close();
    std::cout << "\nWrote complexity table to " << output_complexity_filename << "\n";

}
